### PR TITLE
Implement Fuse Operation Counter

### PR DESCRIPTION
**Pull request by rdpsin**
Signed-off-by: Rishabh Singhvi <rdpsin@amazon.com>

*Issue #, if available:*

*Description of changes:*

*Testing performed:*

```
% curl localhost:6060/metrics | grep image

# HELP soci_fs_image_operation_count_key The count of soci snapshotter operations. Broken down by operation type and image digest.
# TYPE soci_fs_image_operation_count_key gauge
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="file.Getattr"} 0
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="file.Read"} 1463
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="node.Getattr"} 5986
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="node.Getxattr"} 15607
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="node.Listxattr"} 4
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="node.Lookup"} 128940
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="node.Open"} 1935
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="node.Readdir"} 9879
soci_fs_image_operation_count_key{image="sha256:73aeb84f79ad0603eaba572b7d747200d88e7e7e8156d6c8ad64fd3a8f4497d5",operation_type="whiteout.Getattr"} 0
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: fs/fs.go [hunk 1]
```diff
@@ -304,6 +305,11 @@ func (c *sociContext) Init(ctx context.Context, imageRef, indexDigest, imageMani
 		}
 		c.sociIndex = index
 		c.populateImageLayerToSociMapping(index)
+
+		// Create the FUSE operation counter.
+		// Metrics are emitted after a wait time of 60s.
```
Comment (rdpsin) [inline fs/fs.go:310]: Intend to make this configurable.

File: fs/layer/node.go [hunk 2]
```diff
@@ -112,6 +113,47 @@ var fuseOpFailureMetrics = map[string]string{
 	fuseOpWhiteoutGetattr: commonmetrics.FuseWhiteoutGetattrFailureCount,
 }
 
+type FuseOperationCounter interface {
```
Comment (djdongjin) [inline fs/layer/node.go:116]: Can you add comments here? Since it's exported.

File: fs/layer/node.go [hunk 3]
```diff
@@ -112,6 +113,47 @@ var fuseOpFailureMetrics = map[string]string{
...
+	Run()
+}
+
+type defaultFuseOperationCounter struct {
+	opCounts    map[string]*atomic.Int32
+	waitPeriod  time.Duration
+	imageDigest digest.Digest
+}
+
+func NewDefaultOperationCounter(imgDigest digest.Digest, waitPeriod time.Duration) FuseOperationCounter {
```
Comment (djdongjin) [inline fs/layer/node.go:127]: Same here

File: fs/layer/node.go [hunk 4]
```diff
@@ -221,6 +268,10 @@ func (n *node) readdir() ([]fuse.DirEntry, syscall.Errno) {
 	start := time.Now() // set start time
 	defer commonmetrics.MeasureLatencyInMicroseconds(commonmetrics.NodeReaddir, n.fs.layerDigest, start)
 
+	if n.fs.operationCounter != nil {
+		n.fs.operationCounter.Inc(fuseOpReaddir)
+	}
+
```
Comment (djdongjin) [inline fs/layer/node.go:274]: `fuseOpReaddir` is Inc'ed both in `Readdir` and `readdir`. This will double the count right?

Comment (rdpsin) [reply fs/layer/node.go:274]: Yes, will remove it .

File: fs/layer/node.go [hunk 5]
```diff
@@ -112,6 +113,47 @@ var fuseOpFailureMetrics = map[string]string{
...
+	}
+	f.opCounts[fuseOpGetattr] = new(atomic.Int32)
+	f.opCounts[fuseOpGetxattr] = new(atomic.Int32)
+	f.opCounts[fuseOpListxattr] = new(atomic.Int32)
+	f.opCounts[fuseOpLookup] = new(atomic.Int32)
+	f.opCounts[fuseOpOpen] = new(atomic.Int32)
+	f.opCounts[fuseOpReaddir] = new(atomic.Int32)
+	f.opCounts[fuseOpFileRead] = new(atomic.Int32)
+	f.opCounts[fuseOpFileGetattr] = new(atomic.Int32)
+	f.opCounts[fuseOpWhiteoutGetattr] = new(atomic.Int32)
```
Comment (sparr) [inline fs/layer/node.go:141]: This feels like it should have a less repetitive representation. A loop through a slice of operation names would be a bit blunt, but I can't immediately think of a better option.

File: fs/config/config.go [hunk 6]
```diff
@@ -53,15 +53,16 @@ const (
...
+	HTTPCacheType                  string `toml:"http_cache_type"`
+	FSCacheType                    string `toml:"filesystem_cache_type"`
+	ResolveResultEntry             int    `toml:"resolve_result_entry"`
+	Debug                          bool   `toml:"debug"`
+	AllowNoVerification            bool   `toml:"allow_no_verification"`
+	DisableVerification            bool   `toml:"disable_verification"`
+	MaxConcurrency                 int64  `toml:"max_concurrency"`
+	NoPrometheus                   bool   `toml:"no_prometheus"`
+	MountTimeoutSec                int    `toml:"mount_timeout_sec"`
+	FuseMetricsEmitWaitDurationSec int    `toml:"fuse_metrics_emit_wait_duration_sec"`
```
Comment (djdongjin) [inline fs/config/config.go:65]: Should `FuseMetricsEmitWaitDurationSec` be part of `FuseConfig`?

File: fs/layer/node.go [hunk 7]
```diff
@@ -112,6 +113,53 @@ var fuseOpFailureMetrics = map[string]string{
...
 
+// FuseOperationCounter collects number of invocations of the various FUSE implementations and emits them as metrics.
+// Setting `waitPeriod` to be > 0 allows delaying the time when the metrics are emitted.
+type FuseOperationCounter struct {
+	opCounts    map[string]*atomic.Int32
+	waitPeriod  time.Duration
+	imageDigest digest.Digest
+}
+
+var FuseMetricsList = []string{
```
Comment (djdongjin) [inline fs/layer/node.go:124]: `FuseMetricsList` -> `FuseOpList`. Also can you add a comment here or change to not export.

File: fs/layer/node.go [hunk 8]
```diff
@@ -112,6 +113,53 @@ var fuseOpFailureMetrics = map[string]string{
...
+	fuseOpListxattr,
+	fuseOpLookup,
+	fuseOpOpen,
+	fuseOpReaddir,
+	fuseOpFileRead,
+	fuseOpFileGetattr,
+	fuseOpWhiteoutGetattr,
+}
+
+func NewFuseOperationCounter(imgDigest digest.Digest, waitPeriod time.Duration) *FuseOperationCounter {
```
Comment (djdongjin) [inline fs/layer/node.go:136]: can you add brief comment here.

File: fs/layer/node.go [hunk 9]
```diff
@@ -112,6 +113,53 @@ var fuseOpFailureMetrics = map[string]string{
...
+		waitPeriod:  waitPeriod,
+		opCounts:    make(map[string]*atomic.Int32),
+	}
+	for _, m := range FuseMetricsList {
+		f.opCounts[m] = new(atomic.Int32)
+	}
+	return f
+}
+
+func (f *FuseOperationCounter) Inc(op string) {
```
Comment (djdongjin) [inline fs/layer/node.go:148]: can you add brief comment since it's exported.

File: fs/layer/node.go [hunk 10]
```diff
@@ -112,6 +113,53 @@ var fuseOpFailureMetrics = map[string]string{
...
+	operation, ok := f.opCounts[op]
+	if !ok {
+		return
+	}
+	operation.Add(1)
+}
+
+func (f *FuseOperationCounter) Run() {
```
Comment (djdongjin) [inline fs/layer/node.go:156]: can you add brief comment since it's exported.

File: fs/metrics/common/metrics.go [hunk 11]
```diff
@@ -198,3 +211,7 @@ func IncOperationCount(operation string, layer digest.Digest) {
 func AddBytesCount(operation string, layer digest.Digest, bytes int64) {
 	bytesCount.WithLabelValues(operation, layer.String()).Add(float64(bytes))
 }
+
+func AddImageOperationCount(operation string, image digest.Digest, count int32) {
```
Comment (djdongjin) [inline fs/metrics/common/metrics.go:215]: can you add brief comment since it's exported.

File: integration/metrics_test.go [hunk 12]
```diff
@@ -224,6 +226,74 @@ log_fuse_operations = true
...
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			rebootContainerd(t, sh, "", "")
+
+			imgInfo := dockerhub(tc.image)
+			sh.X("ctr", "i", "pull", imgInfo.ref)
+			indexDigest := buildSparseIndex(sh, imgInfo, 0, defaultSpanSize)
```
Comment (djdongjin) [inline integration/metrics_test.go:272]: can you rebase to get the change from https://github.com/awslabs/soci-snapshotter/pull/350

File: integration/metrics_test.go [hunk 13]
```diff
@@ -224,6 +226,74 @@ log_fuse_operations = true
...
+
+			sh.X("soci", "image", "rpull", "--soci-index-digest", indexDigest, imgInfo.ref)
+			sh.XLog("ctr", "run", "--rm", "-d", "--snapshotter=soci", imgInfo.ref, "test", "echo", "hi")
+
+			curlOutput := string(sh.O("curl", tcpMetricsAddress+metricsPath))
+
+			for _, m := range layer.FuseMetricsList {
+				if checkMetricExists(curlOutput, m) {
+					t.Fatalf("didn't expect metric %s but got it", m)
```
Comment (djdongjin) [inline integration/metrics_test.go:281]: nit: -> ("got unexpected metric: %s", m)

File: integration/metrics_test.go [hunk 14]
```diff
@@ -224,6 +226,74 @@ log_fuse_operations = true
...
+				}
+			}
+
+			time.Sleep(10 * time.Second)
+			curlOutput = string(sh.O("curl", tcpMetricsAddress+metricsPath))
+			for _, m := range layer.FuseMetricsList {
+				if !checkMetricExists(curlOutput, m) {
+					t.Fatalf("expected metric %s but didn't get it", m)
```
Comment (djdongjin) [inline integration/metrics_test.go:289]: nit: -> ("missing expected metric: %s", m)

File: fs/config/config.go (hunk 6 above)
Comment (rdpsin) [reply fs/config/config.go:65]: I could go either route. I think it make sense for `FuseConfig` to only contains fields that impacts FUSE. `FuseMetricsEmitWaitDuration` is really part of the filesystem, not the FUSE implementation.

Comment (djdongjin) [reply fs/config/config.go:65]: SGTM, I don't have strong preference either.

**Review by sbuckfelder — APPROVED** [review]
LGTM

**Review by hanyuel — APPROVED** [review]

**Review by Kern-- — APPROVED** [review]

**Pull request merged**